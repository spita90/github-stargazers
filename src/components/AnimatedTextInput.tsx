import React, { useEffect, useRef, useState } from "react";
import {
  Animated,
  Easing,
  Platform,
  StyleProp,
  Text,
  TextInput,
  TextStyle,
  TouchableWithoutFeedback,
  View,
} from "react-native";
import { useTw } from "../theme";

type Props = React.ComponentProps<typeof TextInput> & {
  textStyle?: StyleProp<TextStyle>;
  labelStyle?: StyleProp<TextStyle>;
  label: string;
  editable?: boolean;
};

export const AnimatedTextInput: React.FC<Props> = (props) => {
  const {
    label,
    editable = true,
    value,
    style,
    textStyle,
    labelStyle,
    onBlur,
    onFocus,
    ...restOfProps
  } = props;
  const [tw] = useTw();
  const [isFocused, setIsFocused] = useState(false);
  const inputRef = useRef<TextInput>(null);
  const focusAnim = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.timing(focusAnim, {
      toValue: isFocused || !!value ? 1 : 0,
      duration: 150,
      easing: Easing.bezier(0.4, 0, 0.2, 1),
      useNativeDriver: Platform.OS !== "web",
    }).start();
  }, [focusAnim, isFocused, value]);

  return (
    <View style={style}>
      <TextInput
        style={[
          tw`relative p-6 border-3 border-black rounded-lg text-4`,
          textStyle,
        ]}
        ref={inputRef}
        {...restOfProps}
        value={value}
        editable={editable}
        onBlur={(event) => {
          setIsFocused(false);
          onBlur?.(event);
        }}
        onFocus={(event) => {
          if (editable) setIsFocused(true);
          onFocus?.(event);
        }}
      />
      <TouchableWithoutFeedback onPress={() => inputRef.current?.focus()}>
        <Animated.View
          style={[
            tw`absolute px-[8px] bg-white rounded-sm`,
            {
              transform: [
                {
                  scale: focusAnim.interpolate({
                    inputRange: [0, 1],
                    outputRange: [1, 0.75],
                  }),
                },
                {
                  translateY: focusAnim.interpolate({
                    inputRange: [0, 1],
                    outputRange: [24, -12],
                  }),
                },
                {
                  translateX: focusAnim.interpolate({
                    inputRange: [0, 1],
                    outputRange: [16, 0],
                  }),
                },
              ],
            },
          ]}
        >
          <Text style={[tw`text-4`, labelStyle]}>{label}</Text>
        </Animated.View>
      </TouchableWithoutFeedback>
    </View>
  );
};
