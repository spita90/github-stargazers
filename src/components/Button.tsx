import {
  StyleProp,
  TouchableOpacity,
  View,
  ViewProps,
  ViewStyle,
} from "react-native";
import { useTw } from "../theme";
import { ColorsType } from "../theme/palette";

export interface ButtonProps extends ViewProps {
  onPress?: () => void;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  children?: string | React.ReactNode;
  color?: ColorsType;
}

export const Button: React.FC<ButtonProps> = ({
  onPress,
  disabled,
  style,
  children,
  color,
}: ButtonProps) => {
  const [tw] = useTw();
  return (
    <TouchableOpacity
      activeOpacity={disabled ? 1 : 0.6}
      onPress={() => {
        if (!disabled && onPress) {
          onPress();
        }
      }}
    >
      <View
        style={[
          tw`py-md px-lg rounded-5 flex items-center justify-center bg-${
            color ?? "black"
          }`,
          { opacity: !disabled ? 1 : 0.6 },
          style,
        ]}
      >
        {children}
      </View>
    </TouchableOpacity>
  );
};
