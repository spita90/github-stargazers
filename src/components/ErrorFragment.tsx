import axios from "axios";
import React, { useEffect } from "react";
import { View } from "react-native";
import Toast from "react-native-root-toast";
import { config } from "../config";
import { useTw } from "../theme";
import { DomainError } from "../types";
import { i18n } from "./core/LanguageLoader";
import { Text } from ".";

export interface ErrorPageProps {
  error: Error;
  resetErrorBoundary?: (...args: Array<unknown>) => void;
}

/**
 * To be used for blocking errors
 */
export function ErrorFragment({ error, resetErrorBoundary }: ErrorPageProps) {
  const [tw] = useTw();

  useEffect(() => {
    if (
      axios.isAxiosError(error) &&
      error.code &&
      error.config &&
      error.config.url
    ) {
      error.message = `${error.message} (${error.code} in ${error.config.url}).`;
    }
    if (error && error.hasOwnProperty("message")) {
      const message =
        error instanceof DomainError
          ? i18n.t((error as DomainError).message)
          : error.message;
      Toast.show(message, {
        duration: 5000,
        position: Toast.positions.BOTTOM,
        shadow: true,
        animation: true,
        hideOnPress: true,
        backgroundColor: "red",
        delay: 0,
      });
    }
  }, []);

  return (
    <View
      style={{
        flex: 1,
        flexDirection: "column",
        alignContent: "stretch",
        justifyContent: "center",
      }}
    >
      <Text
        size={"lg"}
        style={{ alignSelf: "center", padding: 10, marginBottom: 20 }}
      >
        {i18n.t("error")}
      </Text>
      <View
        style={tw`absolute right-6 bottom-6 w-[50%] h-[50px] justify-end items-end`}
      >
        <Text size={"sm"} color="black_opacque">
          v{config.version}
        </Text>
      </View>
    </View>
  );
}
