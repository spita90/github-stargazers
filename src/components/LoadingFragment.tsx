import Lottie from "lottie-react-native";
import { Platform, View } from "react-native";
import { useTw } from "../theme";

export const LoadingFragment = () => {
  const [tw] = useTw();

  return (
    <View style={tw`h-full justify-center items-center`}>
      {Platform.OS === "web" && (
        <Lottie
          style={tw`w-[80%]`}
          source={require("../../assets/animations/github.json")}
          loop={true}
          autoPlay
        />
      )}
    </View>
  );
};
