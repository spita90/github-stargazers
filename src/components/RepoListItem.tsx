import { View } from "react-native";
import { TouchableOpacity } from "react-native-gesture-handler";
import { Text } from ".";
import { StarIconSvg } from "../svgs";
import { useTw } from "../theme";
import { GitHubRepo } from "../types";

export interface RepoListItemProps {
  repo: GitHubRepo;
  onPress?: () => void;
}

export function RepoListItem({ repo, onPress }: RepoListItemProps) {
  const [tw] = useTw();

  return (
    <TouchableOpacity
      style={tw`flex flex-row items-center py-sm border-b-[1px] border-grey`}
      onPress={onPress}
    >
      <Text numberOfLines={1}>{repo.name}</Text>
      {repo.stargazers_count > 0 && (
        <View style={tw`flex flex-row items-center`}>
          <Text>{" - "}</Text>
          <StarIconSvg width={15} height={15} color="grey" />
          <Text style={tw`pl-xs`}>{repo.stargazers_count}</Text>
        </View>
      )}
    </TouchableOpacity>
  );
}
