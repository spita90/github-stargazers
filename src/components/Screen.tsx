import {
  SafeAreaView,
  StatusBar,
  StyleProp,
  View,
  ViewStyle,
} from "react-native";
import { useTw } from "../theme";

export interface ScreenProps {
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  padded?: boolean;
  statusBackgroundColor?: string;
}

/**
 * A configurable screen wrapper
 * @param children the screen content
 * @param style the style for the component
 * @param padded redices the children available space
 * @param statusBackgroundColor the background color for the status bar
 */
export function Screen({
  children,
  padded,
  style,
  statusBackgroundColor,
}: ScreenProps) {
  const statusBg = statusBackgroundColor
    ? statusBackgroundColor
    : "transparent";
  const tw = useTw();
  return (
    <SafeAreaView style={[{ flex: 1 }, style]}>
      <StatusBar
        backgroundColor={statusBg}
        barStyle={"dark-content"}
        animated
      />
      <View style={[{ flex: 1 }, tw`${padded ? "px-xs" : ""}`]}>
        {children}
      </View>
    </SafeAreaView>
  );
}
