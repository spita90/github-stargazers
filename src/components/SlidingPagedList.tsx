import { useEffect, useRef } from "react";
import { Animated, Dimensions, Easing, Platform, View } from "react-native";
import { FlatList } from "react-native-gesture-handler";
import { Button, Text } from ".";
import { useTw } from "../theme";
import { ColorsType } from "../theme/palette";
import { i18n } from "./core/LanguageLoader";

export interface SlidingPagedListProps {
  dataMatrix: any[][];
  renderItem: (item: any) => JSX.Element;
  page: number;
  setPage: (page: number) => void;
  title?: string;
  backgroundColor?: ColorsType;
  snapPoints: (string | number)[]; //starting from the top
  initialSnapIndex: number;
  height: number;
  bottomMargin: number;
  visible: boolean;
  setVisible: (visible: boolean) => void;
  sliderRef?: any;
}

export const SlidingPagedList = ({
  sliderRef,
  dataMatrix,
  renderItem,
  page,
  setPage,
  title,
  backgroundColor,
  snapPoints,
  initialSnapIndex,
  height,
  bottomMargin,
  visible,
  setVisible,
}: SlidingPagedListProps) => {
  const [tw] = useTw();

  const canGoBack = page > 0;
  const canGoNext = dataMatrix[page + 1] && dataMatrix[page + 1].length > 0;

  const RESULTS_VIEW_HIDDEN_Y_POS_PX = Dimensions.get("window").height;

  const resultsViewSlideAnim = useRef(
    new Animated.Value(RESULTS_VIEW_HIDDEN_Y_POS_PX)
  ).current;

  useEffect(() => {
    Animated.timing(resultsViewSlideAnim, {
      toValue: visible ? 0 : RESULTS_VIEW_HIDDEN_Y_POS_PX,
      duration: 300,
      easing: Easing.bezier(0.4, 0, 0.2, 1),
      useNativeDriver: Platform.OS !== "web",
    }).start();
  }, [visible]);

  const NavButtons = () => (
    <View style={tw`flex flex-row h-[54px] mt-sm justify-evenly`}>
      <Button
        style={tw`flex flex-1`}
        disabled={!canGoBack}
        onPress={() => {
          if (!canGoBack) return;
          setPage(page - 1);
        }}
      >
        <Text color="white">{i18n.t("prev")}</Text>
      </Button>
      <Button
        style={tw`flex flex-1`}
        disabled={!canGoNext}
        onPress={() => {
          if (!canGoNext) return;
          setPage(page + 1);
        }}
      >
        <Text color={"white"}>{i18n.t("next")}</Text>
      </Button>
    </View>
  );

  const ListHeader = () => (
    <View
      style={[
        tw`flex flex-row items-center justify-between
      pb-md`,
        { backgroundColor: backgroundColor },
      ]}
    >
      {title && (
        <Text textStyle={tw`text-xl`} bold>
          {title}
        </Text>
      )}
      <Button style={tw`px-sm py-xs`} onPress={() => setVisible(false)}>
        <Text color="white">X</Text>
      </Button>
    </View>
  );

  useEffect(() => {
    // sliderRef.current.snapTo(visible ? 0 : snapPoints.length - 1);
  }, [visible]);

  return (
    <Animated.View
      style={[
        tw`absolute w-full`,
        { bottom: -bottomMargin },
        // { height: height, bottom: 0 },
        { transform: [{ translateY: resultsViewSlideAnim }] },
      ]}
    >
      <View
        style={[
          { backgroundColor: backgroundColor },
          { maxHeight: height },
          { width: Dimensions.get("window").width - 40, marginLeft: 20 },
          tw`p-sm
        border-t-3 border-l-3 border-r-3 border-grey 
        rounded-t-lg`,
        ]}
      >
        <ListHeader />
        <FlatList
          // ref={listRef}
          data={dataMatrix[page]}
          showsVerticalScrollIndicator={false}
          renderItem={({ item }) => renderItem(item)}
          keyExtractor={(itm) => itm.id.toString()}
          // estimatedItemSize={estimatedItemSize}
        />
        {(canGoBack || canGoNext) && <NavButtons />}
      </View>
    </Animated.View>
  );
};
