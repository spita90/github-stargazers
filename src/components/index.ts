export { AppLoader } from "./core/AppLoader";
export { LanguageLoader } from "./core/LanguageLoader";
export { ErrorFragment } from "./ErrorFragment";
export { LoadingFragment } from "./LoadingFragment";
export { Text } from "./Text";
export { Button } from "./Button";
export { Screen } from "./Screen";
