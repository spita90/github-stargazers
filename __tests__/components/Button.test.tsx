import { fireEvent, render } from "@testing-library/react-native";
import * as React from "react";
import renderer from "react-test-renderer";
import { Button, Text } from "../../src/components";

it("renders correctly", () => {
  const tree = renderer.create(
    <Button testID="button0">
      <Text>Login</Text>
    </Button>
  );
  expect(tree).toMatchSnapshot();
});

it("actually calls onPress when pressed", () => {
  const mockOnPressFn = jest.fn();
  mockOnPressFn.mockReturnValue({ called: true });

  const { getByTestId } = render(
    <Button testID="button1" onPress={mockOnPressFn}>
      <Text>Login</Text>
    </Button>
  );
  fireEvent.press(getByTestId("button1"));
  expect(mockOnPressFn.mock.results[0].value).toStrictEqual({
    called: true,
  });
});
