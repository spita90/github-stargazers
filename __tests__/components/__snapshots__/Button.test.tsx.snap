// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="button0"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#000",
          "borderRadius": 20,
          "display": "flex",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 16,
        },
        {
          "opacity": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
            },
            {
              "backgroundColor": "transparent",
              "marginTop": undefined,
              "padding": undefined,
            },
          ],
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 16,
            },
            {
              "fontFamily": "Circular-Std-Medium",
            },
            {
              "textAlign": "auto",
            },
            undefined,
            undefined,
            {
              "color": "#000",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
